### PR TITLE
Support gRPC-JSON translate without the google.api.http option.

### LINKED ISSUE #3631: grpc transcoder: support default HTTP binding
*Description*:

Though it is **strongly preferable** in terms of HTTP interface design to explicitly configure mappings, we may support default HTTP binding like this:
https://cloud.google.com/endpoints/docs/grpc/transcoding#where_to_configure_transcoding

https://envoyproxy.slack.com/archives/C78M4KW76/p1528921629000128

### PR DESCRIPTION
**Pull request by lvht**
Description: Support gRPC-JSON translate without the `google.api.http` option.
Risk Level: Low
Testing: Done
Docs Changes: Done
Release Notes: Done

fixes #3631

### CODE REVIEW

**Comment by zuercher** [comment]
Let's get the formatting and docs errors fixed up and then I'll assign this out for review.

@lvht please avoid force-pushing once proper review comments are made -- we'll squash the PR to master when review is done.

/wait

File: source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc [hunk 1]
```diff
@@ -119,9 +119,16 @@ JsonTranscoderConfig::JsonTranscoderConfig(
     }
     for (int i = 0; i < service->method_count(); ++i) {
       auto method = service->method(i);
-      if (!PathMatcherUtility::RegisterByHttpRule(pmb,
-                                                  method->options().GetExtension(google::api::http),
-                                                  ignored_query_parameters, method)) {
+      auto http_rule = method->options().GetExtension(google::api::http);
+
+      if (!http_rule.pattern_case() && proto_config.auto_mapping()) {
+        auto post = "/" + service->file()->package() + "." + service->name() + "/" + method->name();
```
Comment (lizan) [inline source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc]: `service->full_name()` should be good for package + "." + service->name.

File: source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc [hunk 2]
```diff
@@ -119,9 +119,16 @@ JsonTranscoderConfig::JsonTranscoderConfig(
     }
     for (int i = 0; i < service->method_count(); ++i) {
       auto method = service->method(i);
-      if (!PathMatcherUtility::RegisterByHttpRule(pmb,
-                                                  method->options().GetExtension(google::api::http),
-                                                  ignored_query_parameters, method)) {
+      auto http_rule = method->options().GetExtension(google::api::http);
+
+      if (!http_rule.pattern_case() && proto_config.auto_mapping()) {
```
Comment (lizan) [inline source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc]: `if (method->options().HasExtension(google::api::http))`?

File: source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc (hunk 1 above)
Comment (lvht) [reply source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc]: it seems the `service->full_name()` returns `bookstore.Bookstore.DeleteShelf` and we need `bookstore.Bookstore/DeleteShelf`.

File: source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc [hunk 3]
```diff
@@ -119,9 +120,18 @@ JsonTranscoderConfig::JsonTranscoderConfig(
     }
     for (int i = 0; i < service->method_count(); ++i) {
       auto method = service->method(i);
-      if (!PathMatcherUtility::RegisterByHttpRule(pmb,
-                                                  method->options().GetExtension(google::api::http),
-                                                  ignored_query_parameters, method)) {
+
+      HttpRule(http_rule);
```
Comment (lizan) [inline source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc]: nit: `HttpRule rule;`

**Pull request merged**